### PR TITLE
feat: Added global handling middleware

### LINKED ISSUE #39: Feature: Add global exception handling
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

To make exception handling on different layers easier and to prevent code duplication add global handler class for handling exceptions. This class should be located in `Application` layer.

**Describe alternatives you've considered**

**Additional context**
Web Backend
https://learn.microsoft.com/en-us/aspnet/web-api/overview/error-handling/web-api-global-error-handling

### PR DESCRIPTION
**Pull request by Hubertgitck**
Summary
Added global handling middleware.

Details
Global handling middleware which handles all unhandled errors. Middleware is returning ProblemDetails ( according to RFC 7807 ) response which is serialized to Json format. Handler is able to be extended with new custom exceptions which MUST inherti Exception class.

References
[RFC 7807](https://www.rfc-editor.org/rfc/rfc7807)

### CODE REVIEW

File: binder-backend/Binder.Application/Services/Middleware/CustomExceptions/NotFoundException.cs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Binder.Application.Services.Middleware.CustomExceptions;
```
Comment (StraykerPL) [inline binder-backend/Binder.Application/Services/Middleware/CustomExceptions/NotFoundException.cs]: Use casual namespace declaration.

File: binder-backend/Binder.Api/Program.cs [hunk 2]
```diff
@@ -29,11 +29,14 @@ public static void Main(string[] args)
...
             }
 
+            app.UseMiddleware<ExceptionHandlingMiddleware>();
+
             app.UseHttpsRedirection();
             app.UseAuthorization();
             app.MapControllers();
 
             app.Run();
+
```
Comment (StraykerPL) [inline binder-backend/Binder.Api/Program.cs]: Is this line necessary?

File: binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
+﻿using System.Net;
+using System.Text.Json;
+using Binder.Application.Services.Middleware.CustomExceptions;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+
```
Comment (StraykerPL) [inline binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs:7]: Add casual namespace declaration.

File: binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        {
+            await next(context);
+        }
+        catch (Exception e)
+        {
+            ProblemDetails problemDetails = e switch
+            {
+                InvalidOperationException invalidOperationException => new ProblemDetails
+                {
+                    Title = "Invalid operation!",
```
Comment (StraykerPL) [inline binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs]: This strings will be returned to output, make const file for them in Core layer.

File: binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+                    Status = (int)HttpStatusCode.BadRequest,
+                    Detail = invalidOperationException.Message
+                },
+                NotFoundException notFoundException => new ProblemDetails
+                {
+                    Title = "Resource not found!",
```
Comment (StraykerPL) [inline binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs]: Like higher.

File: binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+                    Status = (int)HttpStatusCode.NotFound,
+                    Detail = notFoundException.Message
+                },
+                _ => new ProblemDetails
+                {
+                    Title = "An unexpected error occurred!",
```
Comment (StraykerPL) [inline binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs]: Like higher.

File: binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+                    Status = (int)HttpStatusCode.InternalServerError,
+                    Detail = e.Message
+                },
+            };
+            context.Response.ContentType = "application/problem+json";
```
Comment (StraykerPL) [inline binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs]: Move string to const field.

**Review by StraykerPL — CHANGES_REQUESTED** [review]

**Comment by StraykerPL** [comment]
Please update PR's data: Assignees, Labels and ticket connection.

File: binder-backend/Binder.Api/Program.cs (hunk 2 above)
Comment (Hubertgitck) [reply binder-backend/Binder.Api/Program.cs]: No, i removed it.

File: binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs (hunk 3 above)
Comment (Hubertgitck) [reply binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs:7]: Changed to casual declaration

File: binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs (hunk 4 above)
Comment (Hubertgitck) [reply binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs]: Added file in core layer with const strings for exception titles.

File: binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs (hunk 5 above)
Comment (Hubertgitck) [reply binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs]: Added file in core layer with const strings for exception titles.

File: binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs (hunk 6 above)
Comment (Hubertgitck) [reply binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs]: Added file in core layer with const strings for exception titles.

File: binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs (hunk 7 above)
Comment (Hubertgitck) [reply binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs]: Added file in core layer with const strings for exception titles.

Comment (StraykerPL) [reply binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs]: This comment is marked as resolved, but change was not preformed, please correct the mistake.

**Comment by StraykerPL** [comment]
Please re-request review if work over this PR is finished.

File: binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs (hunk 7 above)
Comment (Hubertgitck) [reply binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs]: Sorry, corrected.

File: binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+                        Detail = notFoundException.Message
+                    },
+                    _ => new ProblemDetails
+                    {
+                        Title = ExceptionConstants.UnexpectedTitle,
+                        Status = (int)HttpStatusCode.InternalServerError,
+                        Detail = e.Message
+                    },
+                };
+                context.Response.ContentType = ExceptionConstants.ProblemJsonType;
```
Comment (StraykerPL) [inline binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs]: In previous comment I marked it to move this value to "const field" not into domain area. Please, next time read comment carefully before performing changes.

**Review by StraykerPL — CHANGES_REQUESTED** [review]

File: binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs (hunk 8 above)
Comment (Hubertgitck) [reply binder-backend/Binder.Application/Services/Middleware/ErrorHandlingMiddleware.cs]: Moved to private const field.

**Review by StraykerPL — APPROVED** [review]

**Pull request merged**